tool/tsh/app_cloud.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// cloudApp is a dummy interface shared between awsApp, azureApp and gcpApp
+type cloudApp interface {
+	dummyCloudAppMethod()
+}
+
+// cloudAppImpl automatically implements cloudApp when inherited
+type cloudAppImpl struct {
+}
+
+func (_ cloudAppImpl) dummyCloudAppMethod() {
+}
+
+// cloudAppInfo contains common information for given cloudApp
+type cloudAppInfo struct {
```

Comment:
Does this is really CloudApp info object ?

Reply:
This holds *common* information about given type of cloud app, i.e. nothing specific to any particular instance (we have existing types for that) but rather metadata: cloud friendly name, callback to create new instance, a method to match `RouteToApp` against that cloud app type. In another language, these could have been expressed as static methods/properties.

We do need this information, but I'm open to suggestions for different names for this type.

Reply:
Removed, this information in now passed in as arguments, one of which makes use of the generic type param.